### PR TITLE
turn on -Werror=unused-function in our Bazel CPU build

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79159
* #79158
* #79155
* #79156
* __->__ #79154
* #79139

Summary:
We also fix any existing issues. Note that we only do this for the CPU
build because nvcc is considered a C++ toolchain but it does not have
the same flag support. Adding flags to the GPU build will cause nvcc
errors.

Test Plan: Built locally, rely on CI to confirm.

Reviewers: malfet

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by seemethere — APPROVED** [review]

**Review by osalpekar — APPROVED** [review]

**Review by seemethere — APPROVED** [review]

File: test/cpp/jit/torch_python_test.cpp [hunk 1]
```diff
@@ -34,29 +34,30 @@ void testEvalModeForLoadedModule() {
   AT_ASSERT(module.attr("dropout").toModule().is_training());
 }
 
-void testSerializationInterop() {
-  if (isSandcastle()) {
-    // The module file to load is not generated in Sandcastle
-    return;
-  }
+// TODO: this test never ran before and is broken.
```
Comment (albanD) [inline test/cpp/jit/torch_python_test.cpp:37]: Why didn't this run?

**Review by albanD — APPROVED** [review]
Code removal :tada: 
Maybe open an issue for the jit test so that it can be looked at.

File: test/cpp/jit/torch_python_test.cpp (hunk 1 above)
Comment (dagitses) [reply test/cpp/jit/torch_python_test.cpp:37]: It was copied into this file like that two years ago: https://github.com/pytorch/pytorch/pull/44795. I commented it out with the same comment that the test invocation was commented out with. I didn't realize it was that old when I kept it in, opening an issue makes sense here. #79275

It looks like when it was copied two years ago it was already dead code.

**Comment by malfet** [comment]
> @pytorchbot revert -m "Breaks bazel build: https://hud.pytorch.org/pytorch/pytorch/commit/67d313a03259be4da7a1d623a9df6791e02248e8 " -c ignoredsignal

Sorry, this was misattribution

**Pull request closed without merging**